### ISSUE TITLE
multiple calls to backward()

### ISSUE THREAD

**Issue by decrispell** [mitsuba-renderer/enoki#94]
I'm confused on how to correctly make multiple calls to ```backward()``` for computing gradients of a computation that has multiple outputs.

In the enoki tests, there are examples for multiple calls to ```forward()```, so I created a similar test for ```backward()``` (copied below).

Based on my interpretation of the [documentation](https://enoki.readthedocs.io/en/master/autodiff.html#custom-forward-and-reverse-mode-traversals), this should work.  However, ```gradient(w)``` is empty, causing a crash when attempting to access element 0 in the first ```assert``` statement.  What am I missing?

```
ENOKI_TEST(test03_mul_bkwd) {
    FloatD w = 4.f, x = 2.f, y = 3.f;
    set_requires_gradient(w);
    set_requires_gradient(x);
    set_requires_gradient(y);

    // A computation that has multiple inputs (w,x,y) and outputs (z1,z2)
    FloatD z1 = x * y;
    FloatD z2 = w * y;

    my_backward(z1, false);
    assert(gradient(w)[0] == 0.f);
    assert(gradient(x)[0] == 3.f);
    assert(gradient(y)[0] == 2.f);

    my_backward(z2);
    assert(gradient(w)[0] == 3.f);
    assert(gradient(x)[0] == 0.f);
    assert(gradient(y)[0] == 4.f);
}
```

**Comment by decrispell** [comment]
Thanks for the speedy response.  As for use-case, I must misunderstand this statement from the doc: 
> However, for a function with n inputs and m outputs, we may want to perform multiple reverse or forward-mode traversals while retaining the computation graph. This is simple to do via an extra argument
`backward(my_variable, free_graph=False)`
`# or`
`forward(my_variable, free_graph=False)`

In any case, thanks to your comments I was able to get my test-case working by removing the `free_graph` argument and resetting the gradients before the second call to `backward()` .  I'll leave it here in case it's helpful for others. Please let me know if I'm still missing something.
```
ENOKI_TEST(test03_mul_bkwd) {
      FloatD w = 4.f, x = 2.f, y = 3.f;
      set_requires_gradient(w);
      set_requires_gradient(x);
      set_requires_gradient(y);

      FloatD z1 = x * y;
      FloatD z2 = w * y;

      my_backward(z1);
      assert(gradient(w)[0] == 0.f);
      assert(gradient(x)[0] == 3.f);
      assert(gradient(y)[0] == 2.f);

      set_gradient(w, 0.f);
      set_gradient(x, 0.f);
      set_gradient(y, 0.f);
      my_backward(z2);
      assert(gradient(w)[0] == 3.f);
      assert(gradient(x)[0] == 0.f);
      assert(gradient(y)[0] == 4.f);
  }
```
